crates/bevy_ecs/src/query/access.rs

```diff
@@ -761,38 +761,61 @@ impl<T: SparseSetIndex> Access<T> {
         self.archetypal.ones().map(T::get_sparse_set_index)
     }
 
-    /// Returns an iterator over the component IDs that this `Access` either
-    /// reads and writes or can't read or write.
+    /// Returns an iterator over the component IDs and their [`ComponentAccessKind`].
     ///
-    /// The returned flag specifies whether the list consists of the components
-    /// that the access *can* read or write (false) or whether the list consists
-    /// of the components that the access *can't* read or write (true).
-    ///
-    /// Because this method depends on internal implementation details of
-    /// `Access`, it's not recommended. Prefer to manage your own lists of
-    /// accessible components if your application needs to do that.
-    #[doc(hidden)]
-    #[deprecated]
-    pub fn component_reads_and_writes(&self) -> (impl Iterator<Item = T> + '_, bool) {
-        (
-            self.component_read_and_writes
-                .ones()
-                .map(T::get_sparse_set_index),
-            self.component_read_and_writes_inverted,
-        )
+    /// Returns `Err(UnboundedAccess)` if the access is unbounded.
+    /// This typically occurs when an [`Access`] is marked as accessing all
+    /// components, and then adding exceptions.
+    pub fn try_iter_component_access(
+        &self,
+    ) -> Result<impl Iterator<Item = ComponentAccessKind<T>> + '_, UnboundedAccess> {
+        if self.component_writes_inverted || self.component_read_and_writes_inverted {
+            return Err(UnboundedAccess);
+        }
+
+        let reads_and_writes = self.component_read_and_writes.ones().map(|index| {
+            let sparse_index = T::get_sparse_set_index(index);
+
+            if self.component_writes.contains(index) {
+                ComponentAccessKind::Exclusive(sparse_index)
+            } else {
+                ComponentAccessKind::Shared(sparse_index)
+            }
+        });
+
+        let archetypical = self
```

Comment:
Chasing it around, the only items `self.archetypal` contains are components explicitly included in a `Query` with `Has<C>`. As such, I'd expect this list to be very short. Additionally, this method was already "wasting" time, since it iterated all `&mut C` items twice (since they're included in `component_writes` _and_ `component_reads_and_writes`). I imagine the net effect here is zero.

My thinking was I could also remove the `archetypal()` method as well, since it's not being used in engine code (or examples). But I decided to just leave them separate just in case to reduce _some_ of the controversy here.